## 45장 프로미스

자바스크립트는 비동기 처리를 위한 하나의 패턴으로 콜백 함수를 사용하지만 콜백 지옥으로 인해 가독성이 나쁘고 비동기 처리 중 발생한 에러의 처리가 힘들며 여러 개의 비동기 처리를 한 번에 처리하는 데도 한계가 있다.

ES6에서는 콜백 패턴이 가진 단점을 보완하는 또 다른 패턴으로 `프로미스`를 도입했다. 프로미스는 비동기 처리 시점을 명확하게 표현할 수 있다는 장점이 있다.

### 1 비동기 처리를 위한 콜백 패턴의 단점

### 1.1 콜백 지옥

```js
// GET 요청을 위한 비동기 함수
const get = (url) => {
  const xhr = new XMLHttpRequest();
  xhr.open('GET', url);
  xhr.send();

  xhr.onload = () => {
    if (xhr.status === 200) {
      // 서버의 응답을 콘솔에 출력한다.
      console.log(JSON.parse(xhr.response));
    } else {
      console.error(`${xhr.status} ${xhr.statusText}`);
    }
  };
};

// id가 1인 post를 취득
get('https://jsonplaceholder.typicode.com/posts/1');
/*
{
    "userId": 1,
    "id": 1,
    "title": "sunt aut facere ... ",
    "body": "quia et suscipit ... "
}
*/
```

비동기 함수를 호출하면 함수 내부의 비동기로 동작하는 코드가 완료되지 않았다 해도 기다리지 않고 즉시 종료된다. 즉, 비동기 함수 내부의 비동기로 동작하는 코드는 비동기 함수가 종료된 이후에 완료된다. 따라서 **비동기 함수 내부의 비동기로 동작하는 코드에서 처리 결과를 외부로 반환하거나 상위 스코프의 변수에 할당하면 기대한 대로 동작하지 않는다.**

예를 들어, `setTimeout` 함수는 비동기 함수다. `setTimeout` 함수를 호출하면 콜백 함수를 호출 스케쥴링한 다음, 타이머 id를 반환하고 즉시 종료된다. 즉, 비동기 함수인 `setTimeout` 함수의 콜백 함수는 `setTimeout` 함수가 종료된 이후에 호출된다. `setTimeout` 함수 내부의 콜백 함수에서 처리 결과를 외부로 반환하거나 상위 스코프의 변수에 할당하면 기대한 대로 동작하지 않는다.

```js
let g = 0;

// 비동기 함수인 setTimeout 함수는 콜백 함수의 처리 결과를 외부로 반환하거나 상위 스코프의 변수에 할당하지 못한다.
setTimeout(() => {
  g = 100;
}, 0);
console.log(g); /// 0
```

더 나은 이해를 위해 앞에 `get()` 함수 예제를 조금 수정해보자! 서버의 응답을 상위 스코프의 변수(todos)에 할당해보자!

```js
let todos;

// GET 요청을 위한 비동기 함수
const get = (url) => {
  const xhr = new XMLHttpRequest();
  xhr.open('GET', url);
  xhr.send();

  xhr.onload = () => {
    console.log('onload 함수 실행! -> 비동기');
    if (xhr.status === 200) {
      // 서버의 응답을 콘솔에 출력한다.
      // console.log(JSON.parse(xhr.response));
      todos = JSON.parse(xhr.response);
    } else {
      console.error(`${xhr.status} ${xhr.statusText}`);
    }
  };
  console.log('get 함수 종료!');
};

// id가 1인 post를 취득
get('https://jsonplaceholder.typicode.com/posts/1');
console.log(todos); // undefined
/*
출력:
get 함수 종료!
undefined
onload 함수 실행! -> 비동기
*/
```

이 또한 기대한 대로 동작하지 않는다. `xhr.onload` 이벤트 핸들러 프로퍼티에 바인딩한 이벤트 핸들러는 `console.log`가 종료한 이후에 호출된다.

그 이유는 `todos`의 변수에 서버의 응답이 할당되기 이전이기 때문이다. 즉, `xhr.onload` 이벤트 핸들러에서 서버의 응답을 상위 스코프의 변수에 할당하면 처리 수선가 보장되지 않는다.

그 이유에 대해 자세하게 알아보자.

1. 비동기 함수 `get`이 호출되면 `get` 함수의 실행 컨텍스트가 생성되고 실행 컨텍스트 스택(콜 스택)에 푸시된다. 이후 함수 코드 실행 과정에서 `xhr.onload` 이벤트 핸들러 프로퍼티에 이벤트 핸들러가 바인딩 된다.

2. `get` 함수가 종료하면 `get` 함수의 실행 컨텍스트가 콜 스택에서 팝되고, 곧바로 `console.log`가 호출된다. 이때 `console.log`의 실행 컨텍스트가 생성되어 실행 컨텍스트 스택에 푸시된다. 이때 만약 `console.log`가 호출되기 직전에 `load` 이벤트가 발생했더라도 `xhr.onload` 이벤트 핸들러 프로퍼티에 바인딩한 이벤트 핸들러는 결코 `console.log` 보다 먼저 실행되지 않는다.

3. 서버로부터 응답이 도착하면 `xhr` 객체에서 `load` 이벤트가 발생한다. 이때 `xhr.onload` 핸들러 프로퍼티에 바인딩한 이벤트 핸들러가 즉시 실행되는 것이 아니라 `xhr.onload` 이벤트 핸들러는 `load` 이벤트가 발생하면 일단 태스크 큐에 저장되어 대기하다가, 콜 스택이 비면 이벤트 루프에 의해 콜 스택으로 푸시되어 실행된다.

4. `xhr.onload` 이벤트 핸들러가 실행되는 시점에 콜 스택이 빈 상태여야 하므로 `console.log`가 종료된 이후에 수행된다. 만약 `get` 함수 이후 `console.log`가 100번 호출된다 해도 `xhr.onload` 이벤트 핸들러는 모든 `console.log`가 종료된 이후에 실행된다.

이처럼 **비동기 함수는 비동기 처리 결과를 외부에 반환할 수 없고, 상위 스코프의 변수에 할당할 수도 없다. 따라서 비동기 함수의 처리 결과(서버의 응답 등)에 대한 후속 처리는 비동기 함수 내부에서 수행해야 한다.** 이때 비동기 함수를 범용적으로 사용하기 위해 비동기 함수에 비동기 처리 결과에 대한 후속 처리를 수행하는 콜백 함수를 전달하는 것이 일반적이다. 필요에 따라 비동기 처리가 성공하면 호출될 콜백 함수와 비동기 처리가 실패하면 호출될 콜백 함수를 전달할 수 있다.

```js
// GET 요청을 위한 비동기 함수
const get = (url, successCallback, failureCallback) => {
  const xhr = new XMLHttpRequest();
  xhr.open('GET', url);
  xhr.send();

  xhr.onload = () => {
    if (xhr.status === 200) {
      // 서버의 응답을 콜백 함수에 인수로 전달하면서 호출하여 응답에 대한 후속 처리를 한다.
      successCallback(JOSN.parse(xhr.response));
    } else {
      // 에러 정보를 콜백 함수에 인수로 전달하면서 호출하여 에러 처리를 한다.
      failureCallback(xhr.status);
    }
  };
};

// id가 1인 post를 취득
// 서버의 응답에 대한 후속 처리를 위한 콜백 함수를 비동기 함수인 get에 전달해야 한다.
get('https://jsonplaceholder.typicode.com/posts/1', console.log, console.error);
```

이처럼 콜백 함수를 통해 비동기 처리 결과에 대한 후속 처리를 수행하는 비동기 함수가 비동기 처리 결과를 가지고 또 다시 비동기 함수를 호출해야 한다면 콜백 함수 호출이 중첩되어 복잡도가 높아지는 현상이 발생하는데, 이르 콜백 지옥이라고 한다. 다음 예제를 보자

```js
// GET 요청을 위한 비동기 함수
const get = (url, callback) => {
  const xhr = new XMLHttpRequest();
  xhr.open('GET', url);
  xhr.send();

  xhr.onload = () => {
    if (xhr.status === 200) {
      // 서버의 응답을 콜백 함수에 전달하면서 호출하여 응답에 대한 후속 처리르 한다.
      callback(JSON.parse(xhr.response));
    } else {
      console.error(`${xhr.status} ${xhr.statusText}`);
    }
  };
};

const url = 'https://jsoplaceholder.typicode.com';

// id가 1인 post의 userId를 취득
get(`${url}/posts/1`, ({ userId }) => {
  console.log(userId); // 1
  // post의 userId를 사용하여 user 정보를 취득
  get(`${url}/users/${userId}`, (userInfo) => {
    console.log(userInfo); // {id: 1, name: "Leanne Graham", username: "Bret", ...}
  });
});
```

### 1.2 에러 처리의 한계

비동기 처리를 위한 콜백 패턴의 문제점 중에서 가장 심각한 것은 에러 처리가 곤란하다는 것이다.

```js
try {
  setTimeout(() => {
    throw new Error('Error');
  }, 1000);
} catch (e) {
  // 에러를 캐치하지 못한다.
  console.error('캐치한 에러', e);
}
```

왜 에러를 처리하지 못할까? 결론부터 말하면 **에러는 호출자 방향으로 전파된다.**

비동기 함수인 `setTimeout`이 호출된 `setTimeout` 함수의 실행 컨텍스트가 생성되어 콜 스택에 푸시되어 실행된다. `setTimeout`은 비동기 함수 이므로 콜백 함수가 호출되는 것을 기다리지 않고 즉시 종료되어 콜 스택에서 제거된다.

이후 타이머가 만료되면 `setTimeout` 함수의 콜백 함수는 태스크 큐로 푸시되고 콜 스택이 비어졌을 때 이벤트 루프에 의해 콜 스택으로 푸시되어 실행된다.

`setTimeout` 함수의 콜백 함수가 실행될 때 `setTimeout` 함수는 이미 콜 스택에서 제거된 상태다. 이것은 `setTimeout` 함수의 콜백 함수를 호출한 것이 `setTimeout` 함수가 아니라는 것을 의미한다.

**에러는 호출자 방향으로 전파된다.**고 했다. 즉, 콜 스택의 아래 방향(실행 중인 실행 컨텍스트가 푸시되기 직전에 푸시된 실행 컨텍스트 방향)으로 전파된다. 하지만 앞에서 살펴본 바와 같이 `setTimeout` 함수의 콜백 함수를 호출한 것은 `setTimeout`이 아니다. 따라서 `setTimeout` 함수의 콜백 함수가 발생시킨 에러는 `catch` 블록에서 캐치되지 않는다.

이러한 비동기 처리를 위한 콜백 패턴의 단점을 보완하기 위해 ES6에서는 **프로미스**<sup>**Promise**</sup>가 도입되었다.

### 2. 프로미스의 생성

promise 생성자 함수를 new 연산자와 함께 호출하면 프로미스(Promise 객체)를 생성한다. ES6에서 도입된 promise는 호스트 객체가 아닌 ECMAScript 사양에 정의된 표준 빌트인 객체다.

promise 생성자 함수는 비동기 처리를 수행할 콜백 함수를 인수로 전달받는데 이 콜백 함수는 `resolve`와 `reject` 함수를 인수로 전달받는다.

```js
// 프로미스 생성
const promise = new Promise((resolve, reject) => {
    // Promise 함수의 콜백 함수 내부에서 비동기 처리를 수행한다.
    if(/* 비동기 처리 성공*/) {
        resolve('result');
    } else {
        reject('failure reason');
    }
})
```

Promise 생성자 함수가 인수로 전달받은 콜백 함수 내부에서 비동기 처리를 수행한다. 이때 비동기 처리가 성공하면 콜백 함수의 인수로 전달받은 resolve 함수를 호출하고, 비동기 처리가 실패하면 reject 함수를 호출한다. 앞에서 살펴본 비동기 함수 get을 프로미스를 사용해 다시 구현해보자.

```js
// GET 요청을 위한 비동기 함수
const promiseGet = (url) => {
  return new Promise((resolve, reject) => {
    const xhr = new XMLHttpRequest();
    xhr.open('GET', url);
    xhr.close();

    xhr.onload = () => {
      if (xhr.status === 200) {
        // 성공적으로 응답을 전달받으면 resolve 함수를 호출한다.
        resolve(JSON.parse(xhr.response));
      } else {
        reject(new Error(xhr.status));
      }
    };
  });
};

// promiseGet 함수는 프로미스를 반환한다.
promiseGet('https://jsonplaceholder.typicode.com/posts/1');
```

프로미스는 비동기 처리가 어떻게 진행되고 있는지 나타내는 **상태**<sup>**state**</sup> 정보를 갖는다.

| 프로미스의 상태정보 | 의미                                 | 상태 변경 조건                   |
| :------------------ | :----------------------------------- | :------------------------------- |
| pending             | 비동기처리가 아직 수행되지 않은 상태 | 프로미스가 생성된 직후 기본 상태 |
| fulfilled           | 비동기 처리가 수행된 상태(성공)      | resolve 함수 호출                |
| reject              | 비동기 처리가 수행된 상태(실패)      | reject 함수 호출                 |

생성된 직후의 프로미스는 기본적으로 `pending` 상태이다.

- 비동기 처리 성공 : resolve 함수를 호출해 프로미스를 `fulfilled` 상태로 변경한다.
- 비동기 처리 실패 : reject 함수를 호출해 프로미스를 `rejected` 상태로 변경한다.

이처럼 **프로미스의 상태는 resolve 또는 reject 함수를 호추하는 것으로 결정된다.**

`fulfilled` 또는 `rejected` 상태를 `settled` 상태라고 한다. `settled` 상태는 `fulfilled` 또는 `rejected` 상태와 상관없이 `pending`이 아닌 상태로 비동기 처리가 수행된 상태를 말한다. 일단 `settled` 상태가 되면 더는 다른 상태로 변화할 수 없다.

### 3. 프로미스의 후속 처리 메서드

프로미스의 비동기 처리 상태가 변화하면 이에 따른 후속 처리를 해야 한다. 이를 위해 프로미스는 후속 메서드 then, catch, finally를 제공한다.

프로미스의 비동기 처리 상태가 변화하면 후속 처리 메서드에 인수로 전달한 콜백 함수가 선택적으로 호출된다.

이 때 후속 처리 메서드의 콜백 함수에 **프로미스의 처리 결과**가 인수로 전달된다. 모든 후속 처리 메서드는 프로미스를 반환하며 비동기로 동작한다.

### 3.1 프로미스의 후속 처리 메서드 - then

### Promise.prototype.then

`then` 메서드는 두 개의 콜백 함수를 인수로 전달 받는다.

- 첫 번째 콜백 함수는 프로미스가 `fulfilled` 상태가 되면 호출된다. 이때 콜백 함수는 프로미스의 비동기 처리 결과를 인수로 전달받는다.
- 두 번째 콜백 함수는 프로미스가 `rejected` 상태가 되면 호출된다. 이때 콜백 함수는 프로미스의 에러를 인수로 전달 받는다.

`then` 메서드는 언제나 프로미스를 반환한다. 만약 `then` 메서드의 콜백 함수가 프로미스를 반환하면 그 프로미스를 그대로 반환하고, 콜백 함수가 프로미스가 아닌 값을 반환하면 그 값을 암묵적으로 `resolve` 또는 `reject`하여 프로미스를 생성한다.

### Promise.prototype.catch

`catch` 메서드의 콜백 함수는 프로미스가 `rejected` 상태인 경우만 호출된다.

### Promise.prototype.finally

`finally` 메서드는 한 개의 콜백 함수를 인수로 전달받는다. `fianlly` 메서드의 콜백 함수는 프로미스의 성공 또는 실패와 상관없이 무조건 한 번 호출된다. `fianlly` 메서드는 프로미스의 상태와 상관없이 공통적으로 수행해야 할 처리 내용이 있을 때 유용하다. `finally` 메서드도 언제나 프로미스를 반환한다.

### 4. 프로미스의 에러 처리

비동기 처리에서 발생한 에러는 `then` 메서드의 두 번째 콜백 함수로 처리할 수 있다.

```js
const wrongUrl = 'https://jsonplaceholder.typicode.com/xxx/1';

// 부적절한 URL이 지정되었기 때문에 에러가 발생한다.
promiseGet(wrongUrl).then(
  (res) => console.log(res),
  (err) => console.error(err)
);
```

비동기 처리에서 발생한 에러는 프로미스의 후속 처리 메서드 `catch`를 사용해 처리할 수 있다.

```js
promiseGet(wrongUrl)
  .then((res) => console.log(res))
  .catch((err) => console.error(err));
```

`catch` 메서드를 호출하면 내부적으로 `then(undefined, onRejected)`을 호출한다. 따라서 위 예제는 내부적으로 다음과 같이 처리된다.

```js
const wrongUrl = 'https://jsonplaceholder.typicode.com/xxx/1';

//부적절한 URL이 지정되었기 때문에 에러가 발생한다.
promiseGet(wrongUrl)
  .then((res) => console.log(res))
  .then(undefined, (err) => console.error(err));
```

단, `then` 메서드의 두 번째 콜백 함수는 첫 번째 콜백 함수에서 발생한 에러를 캐치하지 못하고 코드가 복잡해져서 가독성이 좋지 않다.

```js
promiseGet('https://jsonplaceholder.typicode.com/todos/1').then(
  (res) => console.xxx(res),
  (err) => console.error(err)
); // 두 번째 콜백 함수는 첫 번째 콜백 함수에서 발생한 에러를 캐치하지 못한다.
```

`catch` 메서드를 모든 `then` 메서드를 호출한 이후에 호출하면 비동기 처리에서 발생한 에러(`rejected` 상태)뿐만 아니라 `then` 메서드 내부에서 발생한 에러까지 모두 캐치할 수 있다.

```js
promiseGet('https://jsonplaceholder.typicode.com/todos/1')
  .then((res) => console.log(res))
  .catch((err) => console.error(err));
```

### 5. 프로미스 체이닝

비동기 처리를 위한 콜백 패턴은 콜백 지옥이 발생하는 문제가 있었으며 프로미스는 `then`, `catch`, `finally` 후속 처리 메서드를 통해 콜백 지옥을 해결했다.

```js
const url = 'https://jsonplaceholder.typicode.com';

promiseGet(`${url}/posts/1`)
  .then(({ userId }) => promiseGet(`${url}/users/${userid}`))
  .then((userInfo) => console.log(userInfo))
  .catch((err) => console.error(err));
```

위 예제처럼 `then` → `then` → `catch` 순서로 후속 처리 메서드를 호출했다. 후속 처리 메서드는 언제나 프로미스를 반환하므로 연속적으로 호출할 수 있다. 이를 **프로미스 체이닝**이라 한다.

후속 처리 메서드는 콜백 함수가 반환한 프로미스를 반환한다. 즉, 프로미스도 콜백 패턴을 사용하므로 콜백 함수를 사용하지 않는 것은 아니다.

콜백 패턴은 가독성이 좋지 않다. 따라서 ES8에서는 `async/await`를 통해 해결할 수 있다. `async/await`를 사용하면 프로미스의 후속 처리 메서드 없이 마치 동기 처리처럼 프로미스가 처리 결과를 반환하도록 구현할 수 있다.

```js
const url = 'https://jsonplaceholder.typicode.com';

(async () => {
  // id가 1인 post의 userId를 취득
  const { userId } = await promiseGet(`${url}/posts/1`);

  // 취득한 post의 userId로 user 정보를 취득
  const userInfo = await promiseGet(`${url}/users/${userId}`);

  console.log(userInfo);
})();
```

### 6. 프로미스의 정적 메서드

`Promise`는 주로 생성자 함수로 사용되지만 함수도 객체이므로 메서드를 가질 수 있다. `Promise`는 5가지 정적 메서드를 제공한다.

### 6.1 Promise.resolve / Promise.reject

Promise.resolve와 Promise.reject 메서드는 이미 존재하는 값을 래핑하여 프로미스를 생성하기 위해 사용한다.

#### Promise.resolve

```js
// 배열을 resoulve하는 프로미스를 생성
const resolvedPromise = Promise.resolve([1, 2, 3]);
resolvedPromise.then(console.log);

// 동일하게 동작한다.
const resolvedPromise = enw Promise(resolve  => resolve([1, 2, 3]));
resolvedPromise.then(console.log);
```

#### Promise.reject

```js
// 배열을 resoulve하는 프로미스를 생성
const resolvedPromise = Promise.reject(new Error('Error!'));
resolvedPromise.then(console.log);

// 동일하게 동작한다.
const resolvedPromise = enw Promise((, reject)  => reject(new Error('Error!')));
resolvedPromise.then(console.log);
```

#### Promise.all

`Promise.all` 메서드는 여러 개의 비동기 처리를 모두 병렬처리할 때 사용한다.

```js
const requestData1 = () =>
  new Promise((resolve) => setTimeout(() => resolve(1), 3000));
const requestData2 = () =>
  new Promise((resolve) => setTimeout(() => resolve(2), 2000));
const requestData3 = () =>
  new Promise((resolve) => setTimeout(() => resolve(3), 1000));

// 세 개의 비동기 처리 순차적으로 처리
const res = [];
requestData1()
  .then((data) => {
    res.push(data);
    return requestData2();
  })
  .then((data) => {
    res.push(data);
    return requestData3();
  })
  .then((data) => {
    res.push(data);
    console.log(res);
  })
  .catch(console.error);
```

위 예제는 세 개의 비동기 처리를 순차적으로 처리하기 때문에 6초의 시간이 걸린다.

`Promise.all` 메서드를 사용해 세 개의 비동기 처리를 병렬로 처리해보자.\

```js
const requestData1 = () =>
  new Promise((resolve) => setTimeout(() => resolve(1), 3000));
const requestData2 = () =>
  new Promise((resolve) => setTimeout(() => resolve(2), 2000));
const requestData3 = () =>
  new Promise((resolve) => setTimeout(() => resolve(3), 1000));

// 세 개의 비동기 처리 병렬 처리
Promise.all([requestData1(), requestData2(), requestData3()])
  .then(console.log)
  .catch(console.error);
```

`Promise.all` 메서드는 프로미스를 요소로 갖는 배열 등의 이터러블을 인수로 전달받는다. 전달받은 모든 프로미스가 모두 `fulfilled` 상태가 되면 모든 처리 결과를 배열에 저장해 새로운 프로미스를 반환한다. 이 때 첫 번째 프로미스가 가장 나중에 `fulfilled`가 상태가 되어도 `Promise.all` 메서드는 첫 번째 프로미스가 `resolve`한 처리 결과부터 차례대로 배열에 저장해 그 배열을 `resolve`하는 프로미스를 반환한다. 즉 **처리 순서가 보장된다.**

또한 인수로 전달받은 배열의 모든 프로미스가 모두 `fulfilled` 상태가 되면 종료하며 가장 늦게 `fulfilled` 상태가 되는 프로미스의 처리 시간보다 조금 더 길다. 따라서 위 예제의 경우 가장 늦게 `fulfilled` 상태가 되는 첫 번째 프로미스의 처리 시간인 3초보다 조금 더 소요된다.

그리고 하나라도 `rejected` 상태가 되면 나머지 프로미스가 `fulfilled` 상태가 되는 것을 기다리지 않고 바로 종료한다.

#### Prmise.race

`Promise.race` 메서드는 `Promise.all` 메서드와 동일하게 프로미스를 요소로 갖는 배열 등의 이터러블을 인수로 전달받는다.

`Promise.race`는 `Promise.all` 메서드처럼 모든 프로미스가 `fulfilled` 상태가 되는 것을 기다리는 것이 아니라 가장 먼저 `fulfilled` 상태가 된 프로미스의 처리 결과를 `resolve`하는 새로운 프로미스를 반환한다.

```js
Promise.race([
  new Promise((resolve) => setTimeout(() => resolve(1), 3000)),
  new Promise((resolve) => setTimeout(() => resolve(2), 2000)),
  new Promise((resolve) => setTimeout(() => resolve(3), 1000)),
])
  .then(console.log)
  .catch(console.log);
```

#### Promise.allSettled

`Promise.allSettled` 메서드는 프로미스를 요소로 갖는 배열 등의 이터러블을 인수로 전달받는다. 그리고 전달받은 프로미스가 모두 `settled` 상태(비동기 처리가 수행된 상태, 즉 `fulfilled` 또는 `rejected` 상태)가 되면 처리 결과를 배열로 반환한다.

```js
Promise.allSettled([
  new Promise((resolve) => setTimeout(() => resolve(1), 2000)),
  new Promise((_, reject) =>
    setTimeout(() => reject(new Error('Error!')), 1000)
  ),
]).then(console.log);
```

### 7. 마이크로태스크 큐

```js
setTimeout(() => console.log(1), 0);

Promise.resolve()
  .then(() => console.log(2))
  .then(() => console.log(3));
```

프로미스의 후속 처리 메서드도 비동기로 동작하므로 `1 → 2 → 3`의 순으로 출력될 것처럼 보이지만 `2 → 3 → 1`의 순으로 출력된다. 그 이유는 프로미스의 후속 처리 메서드의 콜백 함수는 태스크 큐가 아니라 마이크로태스크 큐에 저장되기 때문이다.

마이크로태스크 큐는 태스크 큐와 별도의 큐다. 마이크로태스크 큐에는 프로미스의 후속 처리 메서드의 콜백 함수가 일시 저장된다. 그 외의 비동기 함수의 콜백 함수나 이벤트 핸들러는 태스크 큐에 일시 저장된다.

콜백 함수나 이벤트 핸들러를 일시 저장한다는 점에서 태스크 큐와 동일하지만 마이크로태스크 큐는 태스크 큐보다 우선순위가 높다. 즉, 이벤트 루프는 콜 스택이 비면 먼저 마이크로태스크 큐에서 대기하고 있는 함수를 가져와 실행한다. 이후 마이크로태스크 큐가 비면 태스크 큐에서 대기하고 있는 함수를 가져와 실행한다.

### 8. Fetch

`fetch` 함수는 `XMLHttpRequest` 객체와 마찬가지로 HTTP 요청 전송 기능을 제공하는 클라이언트 사이드 Web API다. `fetch` 함수는 `XMLHttpRequest` 객체보다 사용법이 간단하고 프로미스를 지원하기 때문에 비동기 처리를 위한 콜백 패턴의 단점에서 자유롭다.

`fetch` 함수에는 HTTP 요청을 전송할 URL과 HTTP 요청 메서드, HTTP 요청 헤더, 페이로드 등을 설정한 객체를 전달한다.

`const promise = fetch(url [, options])`

**fetch 함수는 HTTP 응답을 나타내는 Response 객체를 래핑한 Promise 객체를 반환한다.** `fetch` 함수로 GET 요청을 전송해보자. `fetch` 함수에 첫 번째 인수로 HTTP 요청을 전송할 URL만 전달하면 GET 요청을 전송한다.

```js
fetch('https://jsonplaceholder.typicode.com/todos/1').then((response) =>
  console.log(response)
);
```

`fetch` 함수는 HTTP 응답을 나타내는 Response 객체를 래핑한 프로미스를 반환하므로 후속 처리 메서드 `then`을 통해 프로미스가 `resolve`한 Response 객체를 전달받을 수 있다. Response 객체는 HTTP 응답을 나타내는 다양한 프로퍼티를 제공한다.

`fetch` 함수가 반환한 프로미스가 래핑하고 있는 MIME 타입이 `application/json`인 HTTP 응답 몸체를 취득하려면 `Response.prototype.json` 메서드를 사용한다. `Response.prototype.json` 메서드는 `Response` 객체에서 HTTP 응답 몸체를 취득하여 역직렬화한다.

```js
fetch('https://jsonplaceholder.typicode.com/todos/1')
  // response는 HTTP 응답을 나타내는 Response 객체다.
  // json 메서드를 사용하여 Response 객체에서 HTTP 응답 몸체를 취득하여 역지렬화한다.
  .then((response) => response.json)
  // json은 역직렬화된 HTTP 응답 몸체다.
  .then((json) => console.log(json));
// {userId: 1, id: 1, title: "delectus aut autem", completed: false}
```

`fetch` 함수를 사용할 때는 에러 처리에 주의해야 한다.

```js
const wrongUrl = 'https://jsonplaceholder.typicode.com/xxx/1';

// 부적절한 URL이 저장되었기 때문에 404 Not Found 에러가 발생한다.
fetch(wrongUrl)
  .then(() => console.log('ok'))
  .catch(() => console.log('error'));
```

부적절한 URL이 지정되었기 때문에 `404 Not Found` 에러가 발생하고 `catch`가 후속 처리 메서드에 의해 'error'를 출력할 것 같지만 'ok'가 출력된다.

이는 `fetch` 함수가 반환하는 프로미스는 기본적으로 `404 Not Found`나 `500 Internal Server Error`와 같은 HTTP 에러가 발생해도 에러를 `reject`하지 않고 불리언 타입의 `ok` 상태를 `false`로 설정한 `Response` 객체를 `resolove`한다. 오프라인 등의 네트워크 장애나 CORS 에러에 의해 요청이 완료되지 못한 경우에만 프로미스를 `reject`한다.

따라서 `fetch` 함수를 사용할 때는 다음과 같이 `fetch` 함수가 반환한 프로미스가 `resolve`한 불리언 타입의 `ok` 상태를 확인해 명시적으로 에러를 처리할 필요가 있다.

```js
const wrongUrl = 'https://jsonplaceholder.typicode.com/xxx/1;';

// 부적절한 URL이 지정되었기 때문에 404 Not Found 에러가 발생한다.
fetch(wrongUrl)
  .then((response) => {
    if (!response.ok) throw new Error(response.statusText);
    return response.json();
  })
  .then((todo) => console.log(todo))
  .catch((err) => console.error(err));
```

참고로 `axios`는 모든 HTTP 에러를 reject하는 프로미스를 반환한다. 따라서 모든 에러를 `catch`에서 처리할 수 있어 편리하다. 또한 `axios`는 인터셉터, 요청 설정 등 `fetch`보다 다양한 기능을 지원한다.

`fetch` 함수를 통해 HTTP 요청을 전송해보자. `fetch` 함수에 첫 번째 인수로 HTTP 요청을 전송할 URL과 두 번째 인수로 HTTP 요청 메서드, HTTP 요청 헤더, 페이로드 등을 설정한 객체를 전달한다.

```js
const request = {
  get(url) {
    return fetch(url);
  },
  post(url, payload) {
    return fetch(url, {
      method: 'POST',
      headers: { 'content-Type': 'application/josn' },
      body: JSON.stringify(payload),
    });
  },
  patch(url, payload) {
    return fetch(url, {
      method: 'PATCH',
      headers: { 'content-Type': 'application/json' },
      body: JSON.stringify(payload),
    });
  },
  delete(url) {
    return fetch(url, { method: 'DELETE' });
  },
};
```

### GET 요청

```js
request
  .get('https://jsonplaceholder.typicode.com/todos/1')
  .then((response) => {
    if (!response.ok) throw new Error(response.setatusText);
    return response.json();
  })
  .then((todos) => console.log(todos))
  .catch((err) => console.error(err));
```

### POST 요청

```js
request
  .post('https://jsonplaceholder.typicode.com/todos/', {
    userId: 1,
    title: 'Javascript',
    completed: false,
  })
  .then((response) => {
    if (!response.ok) throw new Error(response.statusText);
    return response.json();
  })
  .then((todos) => console.log(todos))
  .catch((err) => console.log(err));
```

### PATCH 요청

```js
request
  .patch('https://jsonplaceholder.typicode.com/todos/1', {
    complted: true,
  })
  .then((response) => {
    if (!response.ok) throw new Error(response.statusText);
    return response.json();
  })
  .then((todos) => console.log(todos))
  .catch((err) => console.error(err));
```

### DELETE 요청

```js
request
  .delete('https://jsonplaceholder.typicode.com/todos/1')
  .then((response) => {
    if (!response.ok) throw new Error(response.statusText);
    return response.json();
  })
  .then((todos) => console.log(todos))
  .catch((err) => console.error(err));
```
